optuna/pruners/threshold.py

```diff
@@ -0,0 +1,125 @@
+
+            study = create_study(pruner=ThresholdPruner(lower=0.0))
+            study.optimize(objective_for_lower, n_trials=10)
+
+
+    Args
+        lower:
+            A minimum value which determines whether pruner prunes or not.
+            If an intermediate value is smaller than lower, it prunes.
+        upper:
+            A maximum value which determines whether pruner prunes or not.
+            If an intermediate value is larger than upper, it prunes.
+        n_warmup_steps:
+            Pruning is disabled until the trial exceeds the given number of step.
+        interval_steps:
+            Interval in number of steps between the pruning checks, offset by the warmup steps.
+            If no value has been reported at the time of a pruning check, that particular check
+            will be postponed until a value is reported. Value must be at least 1.
+
+    """
+
+    def __init__(
+        self,
+        lower: Optional[float] = None,
+        upper: Optional[float] = None,
+        n_warmup_steps: int = 0,
+        interval_steps: int = 1,
+    ) -> None:
+
+        if isinstance(lower, bool) or isinstance(upper, bool):
+            raise ValueError("lower and upper should be either None, integers or floating points")
+        if (lower is not None) and (not isinstance(lower, int)) and (not isinstance(lower, float)):
+            raise ValueError("lower should be either None, integer or floating point.")
+        if (upper is not None) and (not isinstance(upper, int)) and (not isinstance(upper, float)):
+            raise ValueError("upper should be either None, integer or floating point.")
+        if lower is None and upper is None:
+            raise ValueError("Either lower or upper must be specified.")
+        if lower is not None and upper is not None and lower > upper:
+            raise ValueError("lower should be smaller than upper.")
+        if n_warmup_steps < 0:
+            raise ValueError(
+                "Number of warmup steps cannot be negative but got {}.".format(n_warmup_steps)
+            )
+        if interval_steps < 1:
+            raise ValueError(
+                "Pruning interval steps must be at least 1 but got {}.".format(interval_steps)
+            )
+
+        self.lower = lower
+        self.upper = upper
+        self._n_warmup_steps = n_warmup_steps
+        self._interval_steps = interval_steps
+
+    def prune(self, study: "optuna.study.Study", trial: optuna.structs.FrozenTrial) -> bool:
+
+        step = trial.last_step
+        if step is None:
+            return False
+
+        n_warmup_steps = self._n_warmup_steps
+        if step <= n_warmup_steps:
+            return False
+
+        if not _is_first_in_interval_step(
+            step, trial.intermediate_values.keys(), n_warmup_steps, self._interval_steps
+        ):
+            return False
+
+        latest_value = trial.intermediate_values[step]
+        if math.isnan(latest_value):
+            return True
```

Comment:
I find this behavior to be useful to the majority of the users (although some might want it the other way) but how about documenting it?